admission/exec: externalize exec admission controller

xmudrii opened the pull request:
**What this PR does / why we need it**:

This PR addresses portion of #66680, by externalizing Exec Admission controller.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Addresses portion of #66680

**Release note**:
```release-note
NONE
```

/assign @deads2k @yue9944882 @sttts

xmudrii committed:
admission/exec: externalize exec admission controller

xmudrii commented on plugin/pkg/admission/exec/admission.go:
```diff
@@ -77,17 +86,20 @@ func NewDenyEscalatingExec() *DenyExec {
+var _ = genericadmissioninitializer.WantsExternalKubeClientSet(&DenyExec{})
+
+// SetInternalKubeClientSet implements the WantsInternalKubeClientSet interface.
+func (d *DenyExec) SetExternalKubeClientSet(client kubernetes.Interface) {
```
the functions are moved up to be closer to type, as requested in other relevant PRs.

sttts commented:
/ok-to-test

yue9944882 commented on plugin/pkg/admission/exec/admission.go:
```diff
@@ -100,24 +112,21 @@ func (d *DenyExec) Validate(a admission.Attributes) (err error) {
 		return admission.NewForbidden(a, err)
 	}
 
-	if pod.Spec.SecurityContext != nil {
```
👍🏻looks really good. am slightly not sure about this line b/c it's possibly changing server's behavior. is it just for a prevention of NPE or any other reason?

yue9944882 replied:
answering myself. removing this line is fine.

yue9944882 commented on plugin/pkg/admission/exec/admission_test.go:
```diff
@@ -146,20 +147,20 @@ func testAdmission(t *testing.T, pod *api.Pod, handler *DenyExec, shouldAccept b
 	hostIPCPod := validPod("hostIPC")
-	hostIPCPod.Spec.SecurityContext = &api.PodSecurityContext{}
-	hostIPCPod.Spec.SecurityContext.HostIPC = true
+	hostIPCPod.Spec.SecurityContext = &corev1.PodSecurityContext{}
```
nits: now we dont need this line, also for L155

xmudrii replied:
Addressed.

I have also removed L53 and L158, because they're very similar. Are those okay to be removed or I should revert changes for those 2?

xmudrii commented:
Addressed comments from @yue9944882.

I have also removed L53 and L158, because they're similar as lines @yue9944882 mentioned. Are those okay to be removed or I should revert changes for those two?

neolit123 commented:
/sig api-machinery

yue9944882 reviewed (approved):
/lgtm

sttts commented on plugin/pkg/admission/exec/admission.go:
```diff
@@ -77,17 +86,20 @@ func NewDenyEscalatingExec() *DenyExec {
-		hostIPC:     false,
-		hostPID:     false,
-		privileged:  true,
+var _ admission.ValidationInterface = &DenyExec{}
```
please move these back directly below the type. They document the intention of the struct.

xmudrii replied:
Done.

sttts commented:
one nit. Otherwise lgtm.

sttts commented on plugin/pkg/admission/exec/admission_test.go:
```diff
@@ -106,7 +105,7 @@ func TestAdmission(t *testing.T) {
 }
 
-func testAdmission(t *testing.T, pod *api.Pod, handler *DenyExec, shouldAccept bool) {
+func testAdmission(t *testing.T, pod *corev1.Pod, handler *DenyExec, shouldAccept bool) {
```
is this change correct? Isn't this the pod being admitted?

yue9944882 replied:
No, it isn't. we don't put any obj into `CONNECT` attribute

wenjiaswe commented:
@yliaog

sttts commented on plugin/pkg/admission/exec/admission.go:
```diff
@@ -62,8 +62,20 @@ type DenyExec struct {
 }
 
 var _ admission.ValidationInterface = &DenyExec{}
+var _ = genericadmissioninitializer.WantsExternalKubeClientSet(&DenyExec{})
```
unify the style here

sttts commented:
/lgtm
/assign @deads2k

xmudrii commented:
/test all

deads2k commented:
/approve
